platform/api/esx/api.go

```diff
@@ -0,0 +1,759 @@
+		options: opts,
+		client: client,
+		ctx:    ctx,
+	}, nil
+}
+
+func getNetworkDevice(net object.NetworkReference) (types.BaseVirtualDevice, error) {
+	backing, err := net.EthernetCardBackingInfo(context.TODO())
+	if err != nil {
+		return nil, err
+	}
+
+	device, err := object.EthernetCardTypes().CreateEthernetCard("e1000", backing)
+	if err != nil {
+		return nil, err
+	}
+
+	return device, nil
+}
+
+func (a *API) buildCloneSpec(baseVM *object.VirtualMachine, folder *object.Folder, network object.NetworkReference, resourcePool *object.ResourcePool, datastore *object.Datastore, ignitionData string) (*types.VirtualMachineCloneSpec, error) {
+	devices, err := baseVM.Device(a.ctx)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't get base VM devices: %v", err)
+	}
+	var card *types.VirtualEthernetCard
+	for _, device := range devices {
+		if c, ok := device.(types.BaseVirtualEthernetCard); ok {
+			card = c.GetVirtualEthernetCard()
+			break
+		}
+	}
+	if card == nil {
+		return nil, fmt.Errorf("No network device found.")
+	}
+
+	netDev, err := getNetworkDevice(network)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't get new network backing device: %v", err)
+	}
+
+	card.Backing = netDev.(types.BaseVirtualEthernetCard).GetVirtualEthernetCard().Backing
+
+	configSpecs := []types.BaseVirtualDeviceConfigSpec{
+		&types.VirtualDeviceConfigSpec{
+			Operation: types.VirtualDeviceConfigSpecOperationEdit,
+			Device:    card,
+		},
+	}
+
+	folderref := folder.Reference()
+	poolref := resourcePool.Reference()
+
+	relocateSpec := types.VirtualMachineRelocateSpec{
+		DeviceChange: configSpecs,
+		Folder:       &folderref,
+		Pool:         &poolref,
+	}
+
+	extraConfigSpec := getExtraConfigSpec(ignitionData)
+
+	cloneSpec := &types.VirtualMachineCloneSpec{
+		Location: relocateSpec,
+		Config:   extraConfigSpec,
+		PowerOn:  false,
+		Template: false,
+	}
+
+	datastoreref := datastore.Reference()
+	cloneSpec.Location.Datastore = &datastoreref
+
+	return cloneSpec, nil
+}
+
+func getExtraConfigSpec(ignitionData string) *types.VirtualMachineConfigSpec {
+	var opts []types.BaseOptionValue
+
+	// It's redundant to set both guestinfo.coreos.config.data &
+	// guestinfo.ovfEnv, however, the guestinfo.ovfEnv is currently being
+	// overwritten whenever the machine boots
+	opts = append(opts, &types.OptionValue{
+		Key:   "guestinfo.coreos.config.data",
+		Value: ignitionData,
+	})
+	opts = append(opts, &types.OptionValue{
+		Key:   "guestinfo.coreos.config.data.encoding",
+		Value: "base64",
+	})
+	ovfEnv := fmt.Sprintf(`<Environment oe:id="" xmlns="http://schemas.dmtf.org/ovf/environment/1" xmlns:oe="https://schemas.dmtf.org/ovf/environment/1" xmlns:xml="http://www.w3.org/XML/1998/namespace" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"><PropertySection><Property oe:key="guestinfo.coreos.config.data" oe:value="%s"/><Property oe:key="guestinfo.coreos.config.data.encoding" oe:value="base64"/></PropertySection></Environment>`, ignitionData)
```

Comment:
ignitionData was base64 encoded at this point. However this function should have been removed, I missed that this was still here. Was a remnant of when I was trying to figure out alternative ways to set the guestinfo properties without them being overwritten by the hypervisor.